src/software/backend/simulation/simulator_robot_singleton.cpp

```diff
@@ -0,0 +1,512 @@
+        .last_applied_acceleration_x       = 0,
+        .last_applied_acceleration_y       = 0,
+        .last_applied_acceleration_angular = 0,
+    };
+    FirmwareRobot_t* firmware_robot = app_firmware_robot_create(
+        chicker, dribbler, &(SimulatorRobotSingleton::getPositionX),
+        &(SimulatorRobotSingleton::getPositionY),
+        &(SimulatorRobotSingleton::getOrientation),
+        &(SimulatorRobotSingleton::getVelocityX),
+        &(SimulatorRobotSingleton::getVelocityY),
+        &(SimulatorRobotSingleton::getVelocityAngular),
+        &(SimulatorRobotSingleton::getBatteryVoltage), front_right_wheel,
+        front_left_wheel, back_right_wheel, back_left_wheel, controller_state,
+        robot_constants);
+
+    return std::unique_ptr<FirmwareRobot_t, FirmwareRobotDeleter>(firmware_robot,
+                                                                  FirmwareRobotDeleter());
+}
+
+float SimulatorRobotSingleton::getPositionX()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getPositionX();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getPositionY()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getPositionY();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getOrientation()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getOrientation();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getVelocityX()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getVelocityX();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getVelocityY()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getVelocityY();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getVelocityAngular()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getVelocityAngular();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
+    return 0.0f;
+}
+
+float SimulatorRobotSingleton::getBatteryVoltage()
+{
+    if (simulator_robot)
+    {
+        return simulator_robot->getBatteryVoltage();
+    }
+    LOG(WARNING)
+        << "SimulatorRobotSingleton called without setting the SimulatorRobot first"
+        << std::endl;
```

Comment:
Similar comment here about constantly re-using this "if robot then do thing else warn" pattern, we should really find a generic solution

Reply:
See above comment for suggestion.

Reply:
done. see above